app/controllers/api/messages_controller.rb

```diff
@@ -0,0 +1,152 @@
+      raise OSM::APIBadUserInput, "No body format was given" if params[:body_format].blank?
+
+      # Extract the arguments
+      if params[:recipient_id]
+        recipient_id = params[:recipient_id].to_i
+        recipient = User.find(recipient_id)
+      elsif params[:recipient]
+        recipient_display_name = params[:recipient]
+        recipient = User.find_by(:display_name => recipient_display_name)
+      else
+        raise OSM::APIBadUserInput, "No recipient was given"
+      end
+      title = params[:title]
+      body = params[:body]
+      body_format = params[:body_format]
+
+      raise OSM::APIRateLimitExceeded if current_user.sent_messages.where("sent_on >= ?", Time.now.utc - 1.hour).count >= current_user.max_messages_per_hour
+
+      @message = Message.new(:sender => current_user,
+                             :recipient => recipient,
+                             :sent_on => Time.now.utc,
+                             :title => title,
+                             :body => body,
+                             :body_format => body_format)
+      @message.save!
+
+      UserMailer.message_notification(@message).deliver_later if @message.notify_recipient?
+
+      # Return a copy of the new message
+      respond_to do |format|
+        format.xml { render :action => :show }
+        format.json { render :action => :show }
+      end
+    end
+
+    # Update read status of a message
+    def update
+      @message = Message.find(params[:id])
+      read_status_idx = %w[true false].index params[:read_status]
+
+      raise OSM::APIBadUserInput, "Invalid value of `read_status` was given" if read_status_idx.nil?
+      raise OSM::APIAccessDenied unless current_user.id == @message.to_user_id
+
+      @message.message_read = read_status_idx.zero?
+      @message.save!
+
+      # Return a copy of the message
+      respond_to do |format|
+        format.xml { render :action => :show }
+        format.json { render :action => :show }
+      end
+    end
+
+    # Delete message by marking it as not visible for the current user
+    def destroy
+      @message = Message.find(params[:id])
+      if current_user.id == @message.from_user_id
+        @message.from_user_visible = false
+      elsif current_user.id == @message.to_user_id
+        @message.to_user_visible = false
+      else
+        raise OSM::APIAccessDenied
+      end
+
+      @message.save!
+
+      # Return a copy of the message
+      respond_to do |format|
+        format.xml { render :action => :show }
+        format.json { render :action => :show }
+      end
+    end
+
+    private
+
+    def show_messages
+      if params[:min_id]
+        @messages = @messages.where("id >= ?", params[:min_id]) unless params[:min_id] == "0"
+        @messages = @messages.order(:id => :asc)
+      else
+        @messages = @messages.where("id <= ?", params[:max_id]) unless params[:max_id].nil? || params[:max_id] == "0"
```

Comment:
Are you sure you want "0" as a special value for max id? Because if treated normally, it means "select messages with id <= 0", that is "select no messages at all".

Reply:
I am open to suggestions.

Maybe it would be enough to simply use missing `max_id` as `get the latest in descending order`

Reply:
Modified to use `from_id` instead of `min_id` and `max_id`, and use absence of `from_id` to retrieve the first batch.